pkg/registration/spoke/managedcluster/claim_reconcile.go

```diff
@@ -46,31 +47,70 @@ func (r *claimReconcile) reconcile(ctx context.Context, cluster *clusterv1.Manag
-func (r *claimReconcile) exposeClaims(ctx context.Context, cluster *clusterv1.ManagedCluster) error {
+func (r *claimReconcile) exposeClaims(_ context.Context, cluster *clusterv1.ManagedCluster) error {
 	var reservedClaims, customClaims []clusterv1.ManagedClusterClaim
+	var clusterClaims []*clusterv1alpha1.ClusterClaim
 
 	// clusterClaim with label `open-cluster-management.io/spoke-only` will not be synced to managedCluster.Status at hub.
-	requirement, _ := labels.NewRequirement(labelCustomizedOnly, selection.DoesNotExist, []string{})
-	selector := labels.NewSelector().Add(*requirement)
-	clusterClaims, err := r.claimLister.List(selector)
+	requirement, err := labels.NewRequirement(labelCustomizedOnly, selection.DoesNotExist, []string{})
 	if err != nil {
-		return fmt.Errorf("unable to list cluster claims: %w", err)
+		return err
+	}
+	selector := labels.NewSelector().Add(*requirement)
+
+	if features.SpokeMutableFeatureGate.Enabled(ocmfeature.ClusterClaim) {
+		clusterClaims, err = r.claimLister.List(selector)
+		if err != nil {
+			return fmt.Errorf("unable to list cluster claims: %w", err)
+		}
 	}
 
 	// check if the cluster claim is one of the reserved claims or has a reserved suffix.
 	// if so, it will be treated as a reserved claim and will always be exposed.
 	reservedClaimNames := sets.New(clusterv1alpha1.ReservedClusterClaimNames[:]...)
 	reservedClaimSuffixes := sets.New(r.reservedClusterClaimSuffixes...)
+
+	// when ClusterProperties feature is not enabled, the informer will not be started and the lister will
+	// return an empty list, so we do not need to check featuregate here.
+	propertiesMap := map[string]*aboutv1alpha1.ClusterProperty{}
+	if features.SpokeMutableFeatureGate.Enabled(ocmfeature.ClusterProperty) {
+		clusterProperties, err := r.aboutLister.List(selector)
+		if err != nil {
+			return fmt.Errorf("unable to list cluster properties: %w", err)
+		}
+
+		for _, property := range clusterProperties {
+			propertiesMap[property.Name] = property
+		}
+	}
+
+	// convert claim to properties
 	for _, clusterClaim := range clusterClaims {
-		managedClusterClaim := clusterv1.ManagedClusterClaim{
-			Name:  clusterClaim.Name,
-			Value: clusterClaim.Spec.Value,
+		// if the claim has the same name with the property, ignore it.
+		if _, ok := propertiesMap[clusterClaim.Name]; ok {
+			continue
 		}
+		propertiesMap[clusterClaim.Name] = &aboutv1alpha1.ClusterProperty{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: clusterClaim.Name,
+			},
+			Spec: aboutv1alpha1.ClusterPropertySpec{
+				Value: clusterClaim.Spec.Value,
+			},
+		}
+	}
 
+	for _, property := range propertiesMap {
+		managedClusterClaim := clusterv1.ManagedClusterClaim{
+			Name:  property.Name,
+			Value: property.Spec.Value,
+		}
 		if matchReservedClaims(reservedClaimNames, reservedClaimSuffixes, managedClusterClaim) {
 			reservedClaims = append(reservedClaims, managedClusterClaim)
+			reservedClaimNames.Insert(property.Name)
```

Comment:
why do we need to change the 4 reservedClaimNames?

Reply:
fixed